core/http/client/src/main/java/org/eclipse/rdf4j/http/client/SharedHttpClientSessionManager.java

```diff
@@ -57,6 +62,39 @@
 
 	private final Map<SPARQLProtocolSession, Boolean> openSessions = new ConcurrentHashMap<>();
 
+	private static final HttpRequestRetryHandler retryHandlerStale = new RetryHandlerStale();
+
+	/**
+	 * Retry handler: closes stale connections and suggests to simply retry the HTTP request once. This seems to be
+	 * necessary for Jetty 9.4.24+
+	 * 
+	 * Other HTTP issues are considered to be more severe, so these requests are not retried.
+	 */
+	private static class RetryHandlerStale implements HttpRequestRetryHandler {
+		private final Logger logger = LoggerFactory.getLogger(RetryHandlerStale.class);
+
+		@Override
+		public boolean retryRequest(IOException ioe, int count, HttpContext context) {
+			// only try this once
+			if (count > 1) {
+				return false;
+			}
+			HttpClientContext clientContext = HttpClientContext.adapt(context);
+			HttpConnection conn = clientContext.getConnection();
+
+			if (conn.isStale()) {
+				try {
+					logger.warn("Closing stale connection");
+					conn.close();
+					return true;
+				} catch (IOException e) {
+					logger.error("Error closing connection", e);
+				}
+			}
+			return false;
```

Comment:
I haven't looked in detail, but is this thread-safe? If (potentially) not, perhaps slap a synchronized block on it to be sure?

Reply:
Good point, I'll have to check if it is

Reply:
added synchronize just to be sure